sklearn/partial_dependence.py

```diff
@@ -0,0 +1,571 @@
+
+from itertools import count
+import numbers
+
+import numpy as np
+from scipy.stats.mstats import mquantiles
+
+from .base import is_classifier, is_regressor
+from .utils.extmath import cartesian
+from .externals.joblib import Parallel, delayed
+from .externals import six
+from .externals.six.moves import map, range, zip
+from .utils import check_array
+from .utils.validation import check_is_fitted
+from .tree._tree import DTYPE
+from .exceptions import NotFittedError
+from .ensemble.gradient_boosting import BaseGradientBoosting
+from .ensemble._gradient_boosting import _partial_dependence_tree
+
+
+__all__ = ['partial_dependence', 'plot_partial_dependence']
+
+
+def _grid_from_X(X, percentiles=(0.05, 0.95), grid_resolution=100):
+    """Generate a grid of points based on the ``percentiles of ``X``.
+
+    The grid is a cartesian product between the columns of Z. The ith column of
+    Z consists in ``grid_resolution`` equally-spaced points between the
+    percentiles of the ith column of X.
+    If ``grid_resolution`` is bigger than the number of unique values in the
+    ith column of X, then those unique values will be used instead.
+
+    Parameters
+    ----------
+    X : ndarray
+        The data
+    percentiles : tuple of floats
+        The percentiles which are used to construct the extreme values of
+        the grid.
+    grid_resolution : int
+        The number of equally spaced points to be placed on the grid for a
+        given column.
+
+    Returns
+    -------
+    grid : ndarray, shape=(n_points, X.shape[1])
+        All data points on the grid. n_points is always ``<= grid_resolution **
+        X.shape[1]``.
+    Z: list of ndarray
+        The values with which the grid has been created. The ndarrays may be of
+        different shape: either (grid_resolution,) or (n_unique_values,).
+    """
+    try:
+        assert len(percentiles) == 2
+    except (AssertionError, TypeError):
+        raise ValueError('percentiles must be a sequence of 2 elements.')
+    if not all(0. <= x <= 1. for x in percentiles):
+        raise ValueError('percentiles values must be in [0, 1].')
+    if percentiles[0] >= percentiles[1]:
+        raise ValueError('percentiles[0] must be strictly less '
+                         'than percentiles[1].')
+
+    if grid_resolution <= 1:
+        raise ValueError('grid_resolution must be strictly greater than 1.')
+
+    values = []
+    for feature in range(X.shape[1]):
+        uniques = np.unique(X[:, feature])
+        if uniques.shape[0] < grid_resolution:
+            # feature has low resolution use unique vals
+            axis = uniques
+        else:
+            # create axis based on percentiles and grid resolution
+            emp_percentiles = mquantiles(X, prob=percentiles, axis=0)
+            if np.allclose(emp_percentiles[0, feature],
+                           emp_percentiles[1, feature]):
+                raise ValueError('percentiles are too close to each other, '
```

Comment:
specify feature

Reply:
I fear that it would probably confuse the user more than anything? It's not really a feature, it's a sanity check that yields a better error message than the one matplotlib would throw.

This only really happens in an edge case where the variable only has one value between the lowest and highest percentiles.